src/validation.cpp

```diff
@@ -1295,36 +1295,41 @@ bool MemPoolAccept::ConsensusScriptChecks(const ATMPArgs& args, Workspace& ws)
     return true;
 }
 
-bool MemPoolAccept::FinalizeSubpackage(const ATMPArgs& args, std::vector<Workspace>& workspaces)
+void MemPoolAccept::FinalizeSubpackage(const ATMPArgs& args)
 {
     AssertLockHeld(cs_main);
     AssertLockHeld(m_pool.cs);
-    const CTransaction& tx = *workspaces.front().m_ptx;
-    const uint256& hash = workspaces.front().m_hash;
-    TxValidationState& state = workspaces.front().m_state;
-    const bool bypass_limits = args.m_bypass_limits;
 
     if (!m_subpackage.m_all_conflicts.empty()) Assume(args.m_allow_replacement);
     // Remove conflicting transactions from the mempool
     for (CTxMemPool::txiter it : m_subpackage.m_all_conflicts)
     {
-        LogDebug(BCLog::MEMPOOL, "replacing mempool tx %s (wtxid=%s, fees=%s, vsize=%s). New tx %s (wtxid=%s, fees=%s, vsize=%s)\n",
+        LogDebug(BCLog::MEMPOOL, "replacing mempool tx %s (wtxid=%s, fees=%s, vsize=%s). ",
                 it->GetTx().GetHash().ToString(),
                 it->GetTx().GetWitnessHash().ToString(),
                 it->GetFee(),
-                it->GetTxSize(),
-                hash.ToString(),
-                tx.GetWitnessHash().ToString(),
-                workspaces[0].m_tx_handle->GetFee(),
-                workspaces[0].m_tx_handle->GetTxSize());
+                it->GetTxSize());
+        FeeFrac feerate = m_subpackage.m_changeset->GetAggregateFeeRate();
+        if (m_subpackage.m_changeset->GetTxCount() == 1) {
+            const CTransaction& tx = m_subpackage.m_changeset->GetTx(0);
+            LogDebug(BCLog::MEMPOOL, "New tx %s (wtxid=%s, fees=%s, vsize=%s)\n",
+                    tx.GetHash().ToString(),
+                    tx.GetWitnessHash().ToString(),
+                    feerate.fee, feerate.size);
+        } else {
+            LogDebug(BCLog::MEMPOOL, "New package with %lu txs, fees=%s, vsize=%s\n",
+                    m_subpackage.m_changeset->GetTxCount(),
+                    feerate.fee,
+                    feerate.size);
+        }
```

Comment:
Should this be just logged once; i.e after replacing the transactions.

Reply:
I tried to make this code analogous to the old code, which repeated information for each transaction.  I'm open to changing the logging but I am always a little afraid of breaking user's scripts or habits if people are used to grepping for a transaction ID and seeing more information? (Though I guess this PR already is breaking the logging for package RBF... so maybe not a big concern.)

Reply:
Yeah will break existing grepping scripts, but I think we will need to do this anyway to improve the logging for package replacements. 

I think we should log the new package or new transaction replacement only once after logging all the replaced transactions:

1. First, log all replaced transactions with their ID, witness, fee, and size.
2. Then:
   - For a single transaction replacement, log its ID, witness, fee, and size.
   - For a package replacement, log that a new package has been received, along with the package's number of transactions, fees, and vsize. Then, log each transaction ID and witness in the package.

This approach preserves the previous logging behavior for individual transactions.
Currently, in this PR, for a single transaction, the logs show the following:

```
2024-10-23T10:26:03.802679Z [httpworker.3] [src/validation.cpp:1310] [FinalizeSubpackage] [mempool] replacing mempool tx 6ae77ff52efae84e0cca125e4b5acfe2f3415298f4f98e850b40f89a2aa5990f (wtxid=57f1edd04c15f6532cdddf4a7c8393f5c5e5584f00003855df8c490053027ac8, fees=10000000, vsize=104).
2024-10-23T10:26:03.802686Z [httpworker.3] [src/validation.cpp:1317] [FinalizeSubpackage] [mempool] New tx e089a6391ae23d41baeed46fb1d8ac04cd0db73e82b4b330275f7d7cafcda33c (wtxid=db89e399619446928fb74a41a5207164beb0fd99972ba147b806a6a24f140e13, fees=490000000, vsize=104)
```

For a package, we see duplicate removal logs, which makes it seem like multiple packages are being added for each replacement:

```
2024-10-23T10:30:28.361089Z [httpworker.3] [src/validation.cpp:1310] [FinalizeSubpackage] [mempool] replacing mempool tx 05ad4c0db1a2fe169acd15bf5d8d1f4fa0ff834b5e8259baab4219f3d20b8991 (wtxid=39676091a39b5f77ee0865c2c52ac4bb8378ff7283fc1668738155ec247c3868, fees=10000, vsize=104).
2024-10-23T10:30:28.361110Z [httpworker.3] [src/validation.cpp:1322] [FinalizeSubpackage] [mempool] New package with 2 txs, fees=50000, vsize=208
2024-10-23T10:30:28.361127Z [httpworker.3] [src/validation.cpp:1310] [FinalizeSubpackage] [mempool] replacing mempool tx 992204e731c78b7afc6e4b17e14adf0f6b93d472fddc4ba184dd86e1899a83d1 (wtxid=05d9f03194089441530d7503a12b7f36086a2c41deac0960e238e426ed59e5f4, fees=10000, vsize=104).
2024-10-23T10:30:28.361139Z [httpworker.3] [src/validation.cpp:1322] [FinalizeSubpackage] [mempool] New package with 2 txs, fees=50000, vsize=208
```

With the idea above, the logs for individual transactions will remain the same, but for packages, the logs will look like this:

```
2024-10-23T11:46:49.459665Z [httpworker.2] [src/validation.cpp:1312] [operator()] [mempool] Replacing mempool tx 05ad4c0db1a2fe169acd15bf5d8d1f4fa0ff834b5e8259baab4219f3d20b8991 (wtxid=39676091a39b5f77ee0865c2c52ac4bb8378ff7283fc1668738155ec247c3868, fees=10000, vsize=104).
2024-10-23T11:46:49.459683Z [httpworker.2] [src/validation.cpp:1312] [operator()] [mempool] Replacing mempool tx 992204e731c78b7afc6e4b17e14adf0f6b93d472fddc4ba184dd86e1899a83d1 (wtxid=05d9f03194089441530d7503a12b7f36086a2c41deac0960e238e426ed59e5f4, fees=10000, vsize=104).
2024-10-23T11:46:49.459696Z [httpworker.2] [src/validation.cpp:1342] [FinalizeSubpackage] [mempool] New package (txs=2, fees=50000, vsize=208)
2024-10-23T11:46:49.459709Z [httpworker.2] [src/validation.cpp:1330] [operator()] [mempool] tx acd7ea806cd261b6c8cdb6d15210dad53daabbe68d5cb99f7df31646a769b69d (wtxid=8a867c2e5aea0d89831d3fe6831535e8cf8ce662ba36b6331fa072babc06f08c)
2024-10-23T11:46:49.459721Z [httpworker.2] [src/validation.cpp:1330] [operator()] [mempool] tx 28deb9d38a4490c3f542f0fd173805eb6c2c8309aa12d5e1669fc71a453d23fe (wtxid=8f8a74f38ed9deada97ef37bb5138b46d16f448da7585e33fc861b4142f764d5)
```

This will make the logs much easier to parse with a script. 
See naive diff implementing this change suggestion https://github.com/ismaelsadeeq/bitcoin/commit/555b05680d8e782fbe0d7b4807d2fce23f512b44

Reply:
Thanks for the ping, sorry for the late response. I agree that a "package hash" on it's own might not be useful for someone interested in "which transactions were replaced by which other transactions". On the other hand, the tracepoints are on purpose only semi-stable as they expose internals. If the internals change, the tracepoints need to change too.

 > which (as of this PR) would now report the package hash along with the transaction ids that are in it?

That could be useful, yes. However, constructing and passing arrays/lists to tracepoints is probably not something we want / can  do (at least not without https://github.com/bitcoin/bitcoin/pull/26593).

One solution could be to update the current tracepoint to a `TRACE8` and have a bool argument indicating if `tx_or_package_hash` is a tx or package hash with `m_subpackage.m_changeset->GetTxCount() == 1`. This wouldn't allow to get data on "which transactions were replaced by which other transactions", but consumers would know that the hash is either referring to a tx or package. While I'm currently using the tracepoint to monitor count, size and fees in replacements, I don't use the txids at the moment. I used them a while ago for [fullrbf.mempool.observer](https://web.archive.org/web/20230213002456/https://fullrbf.mempool.observer/). 

also, CC @virtu who initially worked on these tracepoints. Maybe you have another idea.

Reply:
> One solution could be to update the current tracepoint to a TRACE8 and have a bool argument indicating if tx_or_package_hash is a tx or package hash with m_subpackage.m_changeset->GetTxCount() == 1. 

This could look like e.g. https://github.com/0xB10C/bitcoin/commit/2eb1410e774636040d19d3baf5e22b2e4a8fbfd2

Reply:
Thanks! I've included your commit (and will mark this as resolved).